### PR TITLE
Adding a region selector to the Deploy SAM Application wizard

### PR DESCRIPTION
**Pull request by bryceitoc9**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Infrastructure (change which improves the lifecycle management (CI/CD, build, package, deploy, lint, etc) of the application, but does not change functionality.)
- [ ] Technical debt (change which improves the maintainability of the codebase, but does not change functionality)
- [ ] Testing (change which modifies or adds test coverage, but does not change functionality)
- [ ] Documentation (change which modifies or adds documentation, but does not change functionality)

## Description

Adds a region selector to the Deploy SAM Application wizard.
TODO:
- [ ] ~Add S3 region checker~ Moved to https://github.com/aws/aws-toolkit-vscode/issues/354
- [x] Add tests
- [x] Change the failure text that's printed to the output channel to include the region flag (this is a direct dump from `stderr`, can we trust that this always ends with the command?
  - Error now reads as
```
Failed to deploy SAM application. Error while deploying: Error: sam deploy encountered an error: Failed to create/update the stack. Run the following command
to fetch the list of events leading up to the failure
aws cloudformation describe-stack-events --stack-name lkajslkjas --region ap-northeast-1
```

## Motivation and Context

This allows people to deploy to regions other than the one tied to their profile in `~/.aws/config`

## Related Issue(s)

https://github.com/aws/aws-toolkit-vscode/issues/315

## Testing

Manual runs--confirmed that this succeeds when using an S3 bucket in the selected region and fails when selecting a different region.

## Checklist

- [x] I have read the **README** document
- [x] I have read the **CONTRIBUTING** document
- [x] My code follows the code style of this project
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/lambda/wizards/samDeployWizard.ts [hunk 1]
```diff
@@ -195,27 +197,35 @@ export class SamDeployWizard extends MultiStepWizard<SamDeployWizardResponse> {
...
         }
     }
 
+    private readonly REGION: WizardStep = async () => {
+        this.response.region = await this.contextCommands.onCommandSelectRegion()
+
+        return this.response.region ? this.TEMPLATE : undefined
+    }
+
     private readonly TEMPLATE: WizardStep = async () => {
```
Comment (awschristou) [inline src/lambda/wizards/samDeployWizard.ts]: `TEMPLATE` needs to go "back" to `REGION`
It also needs a back button now.

File: src/shared/defaultAwsContextCommands.ts [hunk 2]
```diff
@@ -122,6 +122,11 @@ export class DefaultAWSContextCommands {
         }
     }
 
+    public async onCommandSelectRegion(filter: (region: RegionInfo) => boolean = () => true
+        ): Promise<string | undefined> {
+        return await this.promptForFilteredRegion(filter)
```
Comment (bryceitoc9) [inline src/shared/defaultAwsContextCommands.ts]: Now that I think about it, I can probably just expose promptForFilteredRegion

Comment (bryceitoc9) [reply src/shared/defaultAwsContextCommands.ts]: Removed.

File: src/lambda/wizards/samDeployWizard.ts (hunk 1 above)
Comment (bryceitoc9) [reply src/lambda/wizards/samDeployWizard.ts]: Implemented (the first revision was before merging back button changes)

**Comment by bryceitoc9** [comment]
re: todo item: `Change the failure text that's printed to the output channel to include the region flag (this is a direct dump from stderr, can we trust that this always ends with the command?`, the error in question comes from the AWS CLI itself: https://github.com/aws/aws-cli/blob/4ff0cbacbac69a21d4dd701921fe0759cf7852ed/awscli/customizations/cloudformation/exceptions.py#L42

File: src/shared/ui/picker.ts [hunk 3]
```diff
@@ -43,6 +44,7 @@ export function createQuickPick<T extends vscode.QuickPickItem>({
     if (options) {
         picker.title = options.title
         picker.placeholder = options.placeHolder
+        picker.value = options.value || ''
```
Comment (awschristou) [inline src/shared/ui/picker.ts]: Please add corresponding test(s) to picker.test.ts for this new field

File: src/lambda/lambdaTreeDataProvider.ts [hunk 4]
```diff
@@ -98,7 +98,8 @@ export class LambdaTreeDataProvider implements vscode.TreeDataProvider<AWSTreeNo
 
         registerCommand({
             command: 'aws.deploySamApplication',
-            callback: async () => await deploySamApplication({outputChannel: this.lambdaOutputChannel})
+            callback: async () => await deploySamApplication({outputChannel: this.lambdaOutputChannel,
```
Comment (mpiroc) [inline src/lambda/lambdaTreeDataProvider.ts]: Nit: Indentation

File: src/lambda/wizards/samDeployWizard.ts [hunk 5]
```diff
@@ -107,14 +113,67 @@ class DefaultSamDeployWizardContext implements SamDeployWizardContext {
         return choices[0].uri
     }
 
+    public async promptUserForRegion(regionProvider: RegionProvider,
```
Comment (mpiroc) [inline src/lambda/wizards/samDeployWizard.ts]: Nit: Indentation

File: src/lambda/wizards/samDeployWizard.ts [hunk 6]
```diff
@@ -107,14 +113,67 @@ class DefaultSamDeployWizardContext implements SamDeployWizardContext {
...
+                                     initialRegionCode: string): Promise<string | undefined> {
+        const logger = getLogger()
+        const regionData = await regionProvider.getRegionData()
+
+        const quickPick = picker.createQuickPick<vscode.QuickPickItem>({
+            options: {
+                title: localize('AWS.samcli.deploy.region.prompt',
```
Comment (mpiroc) [inline src/lambda/wizards/samDeployWizard.ts]: Nit: Indentation

File: src/lambda/wizards/samDeployWizard.ts [hunk 7]
```diff
@@ -107,14 +113,67 @@ class DefaultSamDeployWizardContext implements SamDeployWizardContext {
...
+                matchOnDetail: true,
+                ignoreFocusOut: true,
+            },
+            items: regionData.map(r => ({
+                label: r.regionName,
+                detail: r.regionCode,
+                // this is the only way to get this to show on going back
+                // this will make it so it always shows even when searching for something else
+                alwaysShow: r.regionCode === initialRegionCode,
+                description: r.regionCode === initialRegionCode ? localize('AWS.samcli.deploy.region.previousRegion',
```
Comment (mpiroc) [inline src/lambda/wizards/samDeployWizard.ts]: Nit: Indentation

File: src/lambda/wizards/samDeployWizard.ts [hunk 8]
```diff
@@ -125,6 +184,9 @@ class DefaultSamDeployWizardContext implements SamDeployWizardContext {
                     'Enter the AWS S3 bucket to which your code should be deployed'
                 ),
                 ignoreFocusOut: true,
+                prompt: localize('AWS.samcli.deploy.s3Bucket.region',
```
Comment (mpiroc) [inline src/lambda/wizards/samDeployWizard.ts]: Nit: Indentation

File: src/lambda/wizards/samDeployWizard.ts [hunk 9]
```diff
@@ -113,33 +113,38 @@ class DefaultSamDeployWizardContext implements SamDeployWizardContext {
...
-                                                                : ''
-            })),
-            buttons: [
-                vscode.QuickInputButtons.Back
-            ],
-        })
+        const quickPick = picker.createQuickPick<vscode.QuickPickItem>(
+            {
+                options: {
+                    title: localize('AWS.samcli.deploy.region.prompt',
```
Comment (mpiroc) [inline src/lambda/wizards/samDeployWizard.ts]: It's fine to include the braces on the same line like in your previous iteration. This comment was about the `localize` call, which should be:

```typescript
title: localize(
    'AWS.samcli.deploy.region.prompt',
    'Which AWS Region would you like to deploy to?'
)
```

Comment (bryceitoc9) [reply src/lambda/wizards/samDeployWizard.ts]: Done

**Pull request merged**